gammapy/cube/psf_kernel.py

```diff
@@ -0,0 +1,310 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+from __future__ import absolute_import, division, print_function, unicode_literals
+import numpy as np
+from astropy.coordinates import Angle
+from astropy.coordinates.angle_utilities import angular_separation
+import astropy.units as u
+from astropy.units import Quantity
+from astropy.convolution import convolve_fft
+from ..maps import Map, WcsGeom
+from ..image.models.gauss import Gauss2DPDF
+from ..irf import TablePSF
+
+__all__ = [
+   'table_psf_to_kernel_map',
+    'energy_dependent_table_psf_to_kernel_map',
+    'PSFKernel',
+]
+
+
+def table_psf_to_kernel_map(table_psf, geom, normalize=True, factor=4):
+    """Compute a PSF kernel on a given MapGeom.
+    If the MapGeom is not an image, the same kernel will be present on all axes.
+
+    The PSF is estimated by oversampling defined by a given factor.
+
+    Parameters
+    ----------
+    table_psf : `~gammapy.irf.TablePSF`
+        the input table PSF
+    geom : `~gammapy.maps.MapGeom`
+        the target geometry. The PSF kernel will be centered on the spatial center.
+    normalize : bool
+        normalize the PSF kernel (per energy)
+    factor : int
+        the oversample factor to compute the PSF
+    """
+    # First upsample spatial geom
+    upsampled_image_geom = geom.to_image().upsample(factor)
+
+    # get center coordinate
+    center_coord = upsampled_image_geom.center_coord * u.deg
+
+    # get coordinates
+    map_c = upsampled_image_geom.get_coord()
+
+    # get solid angles
+    solid_angles = upsampled_image_geom.solid_angle()
+
+    # compute distances to map center
+    rads = angular_separation(center_coord[0], center_coord[1], map_c.lon * u.deg, map_c.lat * u.deg)
+
+    vals = table_psf.evaluate(rad=rads).reshape(solid_angles.shape) * solid_angles
+
+    # Create map
+    kernel_map = Map.from_geom(geom=upsampled_image_geom.to_cube(axes=geom.axes),
+                               unit='')
+
+    # loop over images and fill map
+    for img, idx in kernel_map.iter_by_image():
+        img += vals.value
```

Comment:
Normalization of the kernels could be added here to avoid the second Python loop at the end.